### PR TITLE
Tidy defaults & test object constructor

### PR DESCRIPTION
**Pull request by kingo55**
This is regarding the #61 and cleaning up the constructor.

While reading through, I thought I'd tidy as I go to illustrate. There's a lot of small tweaks we could do. More details in the comments below:

### CODE REVIEW

File: lib/mojito.js [hunk 1]
```diff
@@ -1,15 +1,18 @@
 Mojito = (function () {
-    // Enable debugging through the console
-    var Options = {
+    var Defaults = {
```
Comment (kingo55) [inline lib/mojito.js]: I think `Defaults` or `defaultOptions` is a more descriptive choice of variable name here.

File: lib/mojito.js [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
-        defaultWaitTimeout: 2000
+        defaultWaitTimeout: 2000,
+        storageAdapter: {
+            onExposure: function () {},
+            onRecipeFailure: function () {}
+        },
+        decisionAdapter: function (test) {
+            return test.getSeededRNGRandom();
+        },
+        cookiePrefix: '_mojito'
```
Comment (kingo55) [inline lib/mojito.js]: We should move these to the defaults for cleanliness.

File: lib/mojito.js [hunk 3]
```diff
@@ -23,88 +26,48 @@ Mojito = (function () {
...
-        var cookiePrefix = '_mojito',
-            p;
-
         var constructor = function (options) {
+            var p;
+
             this.options = Utils.extend({
                 id: null,
-                name: null,
+                name: options.id,
```
Comment (kingo55) [inline lib/mojito.js:35]: We can default the name to the test's ID if it doesn't exist. I don't see why it's necessary to throw an error for this when some users may not need a test name variable exposed.

File: lib/mojito.js [hunk 4]
```diff
@@ -23,88 +26,48 @@ Mojito = (function () {
...
                 recipes: null,
+                state: OBJECT_STATES.STAGING,
```
Comment (kingo55) [inline lib/mojito.js:37]: Default experiments to `staging` mode if state is not set.

File: lib/mojito.js [hunk 5]
```diff
@@ -23,88 +26,48 @@ Mojito = (function () {
...
+            this.options.decisionAdapter = this.options.options.decisionAdapter;
 
             // Check params
             if (this.options.id === null) throw new Error('An id for this test must be specified.');
-            if (this.options.name === null) throw new Error('A name for this test must be specified.');
-            if (this.options.recipes === null) throw new Error('Recipes must be specified for this test.');
-            if (this.recipes.length < 2) throw new Error('You must specify at least 2 recipes for a test.');
-            if (!this.options.trigger||!Utils.isFunction(this.options.trigger)) throw new Error('Specify a trigger function for this test.');
+            if (this.recipes && this.recipes.length < 2) throw new Error('You must specify at least 2 recipes for a test.');
             if (this.options.sampleRate > 1) throw new Error('Sample rate must be less or equal to 1');
```
Comment (kingo55) [inline lib/mojito.js:56]: I think we can simplify these checks.

File: lib/mojito.js [hunk 6]
```diff
@@ -23,88 +26,48 @@ Mojito = (function () {
...
-            // Decision adapter definition priority: 1) Test adapter; 2) Global adapter; 3) Default adapter
-            if (!this.options.decisionAdapter) {
-                if (this.options.options && this.options.options.decisionAdapter) {
-                    this.options.decisionAdapter = this.options.options.decisionAdapter;
-                } else {
-                    this.options.decisionAdapter = function (testObject) {
-                        return testObject.getSeededRNGRandom();
-                    };
-                }
-            }
```
Comment (kingo55) [inline lib/mojito.js]: All of this can be removed because we added these to the `defaults` section.

File: lib/mojito.js [hunk 7]
```diff
@@ -1,15 +1,19 @@
...
+        defaultWaitTimeout: 2000,
+        storageAdapter: {
+            onExposure: function () {},
+            onRecipeFailure: function () {}
+        },
+        decisionAdapter: function (test) {
+            return test.getSeededRNGRandom();
+        },
+        cookiePrefix: '_mojito',
+        excluded: !navigator.cookieEnabled
```
Comment (kingo55) [inline lib/mojito.js:16]: We can keep all the default settings in one place

**Review by allmywant — APPROVED** [review]

**Review by dapperdrop — APPROVED** [review]

**Pull request merged**